arangod/Aql/Functions.cpp

```diff
@@ -2582,8 +2583,55 @@ AqlValue functions::Substring(ExpressionContext* ctx, AstNode const&,
 
   return AqlValue(utf8);
 }
-////////////////////////////////////////////////////////////////////////////////
-////////////////////////////////////////////////////////////////////////////////
+
+AqlValue functions::SubstringBytes(ExpressionContext* ctx, AstNode const& node,
+                                   VPackFunctionParametersView parameters) {
+  AqlValue const& value = extractFunctionParameterValue(parameters, 0);
+
+  if (!value.isString()) {
+    registerWarning(ctx, getFunctionName(node).data(), TRI_ERROR_BAD_PARAMETER);
+    return AqlValue{AqlValueHintNull{}};
+  }
+
+  int32_t const start = static_cast<int32_t>(
+      extractFunctionParameterValue(parameters, 1).toInt64());
+  int32_t const end = static_cast<int32_t>(
+      extractFunctionParameterValue(parameters, 2).toInt64());
+
+  if (end < 0 || start < 0 || end < start) {
+    registerWarning(ctx, getFunctionName(node).data(), TRI_ERROR_BAD_PARAMETER);
+    return AqlValue{AqlValueHintNull{}};
+  }
+
+  auto const str = value.slice().stringView();
+
+  if (str.size() < uint32_t(end)) {
+    return AqlValue{velocypack::Slice::emptyStringSlice()};
+  }
+
+  auto const subStr = str.substr(start, end - start);
+
+  auto validate = [](std::string_view str) noexcept {
+    auto* begin = reinterpret_cast<irs::byte_type const*>(str.data());
+    auto* end = begin + str.size();
+
+    while (begin != end) {
+      const auto c = irs::utf8_utils::next_checked(begin, end);
+
+      if (irs::utf8_utils::INVALID_CODE_POINT == c) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  if (!validate(subStr)) {
```

Comment:
Do we want to always make it?
Maybe add some unchecked options?

Reply:
I fear yes, because AQL doesn't support binary data now :-(